src/mlpack/core/tree/rectangle_tree/rectangle_tree_impl.hpp

```diff
@@ -1409,41 +1409,72 @@ void RectangleTree<MetricType, StatisticType, MatType, SplitType, DescentType,
     parent = NULL;
   }
 
-  ar & BOOST_SERIALIZATION_NVP(maxNumChildren);
-  ar & BOOST_SERIALIZATION_NVP(minNumChildren);
-  ar & BOOST_SERIALIZATION_NVP(numChildren);
-  if (Archive::is_loading::value)
+  bool hasParent = (parent != NULL);
+
+  ar & CEREAL_NVP(maxNumChildren);
+  ar & CEREAL_NVP(minNumChildren);
+  ar & CEREAL_NVP(numChildren);
+  if (cereal::is_loading<Archive>())
     children.resize(maxNumChildren + 1);
 
-  ar & BOOST_SERIALIZATION_NVP(begin);
-  ar & BOOST_SERIALIZATION_NVP(count);
-  ar & BOOST_SERIALIZATION_NVP(numDescendants);
-  ar & BOOST_SERIALIZATION_NVP(maxLeafSize);
-  ar & BOOST_SERIALIZATION_NVP(minLeafSize);
-  ar & BOOST_SERIALIZATION_NVP(bound);
-  ar & BOOST_SERIALIZATION_NVP(stat);
-  ar & BOOST_SERIALIZATION_NVP(parentDistance);
-  ar & BOOST_SERIALIZATION_NVP(dataset);
-  ar & BOOST_SERIALIZATION_NVP(ownsDataset);
+  ar & CEREAL_NVP(begin);
+  ar & CEREAL_NVP(count);
+  ar & CEREAL_NVP(numDescendants);
+  ar & CEREAL_NVP(maxLeafSize);
+  ar & CEREAL_NVP(minLeafSize);
+  ar & CEREAL_NVP(bound);
+  ar & CEREAL_NVP(stat);
+  ar & CEREAL_NVP(parentDistance);
+  ar & CEREAL_NVP(hasParent);
+
+  if (!hasParent)
+  {
+    MatType* datasetTemp = const_cast<MatType*>(dataset);
+    ar & CEREAL_POINTER(datasetTemp);
+    if(cereal::is_loading<Archive>())
+    {
+      dataset = datasetTemp;
+    }
```

Comment:
Same comment here; I think if we use `MatType*&` then this block isn't necessary.